velox/common/memory/SharedArbitrator.cpp

```diff
@@ -728,72 +820,105 @@ std::string SharedArbitrator::toString() const {
-    SharedArbitrator* arbitrator)
-    : requestor_(nullptr),
-      arbitrator_(arbitrator),
-      startTime_(std::chrono::steady_clock::now()),
-      arbitrationCtx_(requestor_) {
-  VELOX_CHECK_NOT_NULL(arbitrator_);
-  arbitrator_->startArbitration("Wait for arbitration");
-}
-
-SharedArbitrator::ScopedArbitration::ScopedArbitration(
-    MemoryPool* requestor,
-    SharedArbitrator* arbitrator)
-    : requestor_(requestor),
+    SharedArbitrator* arbitrator,
+    ArbitrationOperation* operation)
+    : operation_(operation),
       arbitrator_(arbitrator),
-      startTime_(std::chrono::steady_clock::now()),
-      arbitrationCtx_(requestor_) {
-  VELOX_CHECK_NOT_NULL(requestor_);
+      arbitrationCtx_(operation->requestPool),
+      startTime_(std::chrono::steady_clock::now()) {
   VELOX_CHECK_NOT_NULL(arbitrator_);
-  requestor_->enterArbitration();
-  arbitrator_->startArbitration(fmt::format(
-      "Wait for arbitration, requestor: {}[{}]",
-      requestor_->name(),
-      requestor_->root()->name()));
-  if (arbitrator_->arbitrationStateCheckCb_ != nullptr) {
-    arbitrator_->arbitrationStateCheckCb_(*requestor_);
+  VELOX_CHECK_NOT_NULL(operation_);
+  operation_->enterArbitration();
+  if (arbitrator_->arbitrationStateCheckCb_ != nullptr &&
+      operation_->requestPool != nullptr) {
+    arbitrator_->arbitrationStateCheckCb_(*operation_->requestPool);
   }
+  arbitrator_->startArbitration(operation_);
 }
 
 SharedArbitrator::ScopedArbitration::~ScopedArbitration() {
-  if (requestor_ != nullptr) {
-    requestor_->leaveArbitration();
-  }
+  operation_->leaveArbitration();
+  arbitrator_->finishArbitration(operation_);
+
+  // Report arbitration operation stats.
   const auto arbitrationTimeUs =
       std::chrono::duration_cast<std::chrono::microseconds>(
-          std::chrono::steady_clock::now() - startTime_)
+          std::chrono::steady_clock::now() - operation_->startTime)
           .count();
   RECORD_HISTOGRAM_METRIC_VALUE(
       kMetricArbitratorArbitrationTimeMs, arbitrationTimeUs / 1'000);
   addThreadLocalRuntimeStat(
       kMemoryArbitrationWallNanos,
       RuntimeCounter(arbitrationTimeUs * 1'000, RuntimeCounter::Unit::kNanos));
+  if (operation_->localArbitrationQueueTimeUs != 0) {
+    addThreadLocalRuntimeStat(
+        kLocalArbitrationQueueWallNanos,
+        RuntimeCounter(
+            operation_->localArbitrationQueueTimeUs * 1'000,
+            RuntimeCounter::Unit::kNanos));
+  }
+  if (operation_->localArbitrationLockWaitTimeUs != 0) {
+    addThreadLocalRuntimeStat(
+        kLocalArbitrationLockWaitWallNanos,
+        RuntimeCounter(
+            operation_->localArbitrationLockWaitTimeUs * 1'000,
+            RuntimeCounter::Unit::kNanos));
+  }
+  if (operation_->globalArbitrationLockWaitTimeUs != 0) {
+    addThreadLocalRuntimeStat(
+        kGlobalArbitrationLockWaitWallNanos,
+        RuntimeCounter(
+            operation_->globalArbitrationLockWaitTimeUs * 1'000,
```

Comment:
By putting global in a queue the lock wait time can be further decomposed as 
lock wait time + queue wait time, like local one.